R/redist_alg_helpers.R

```diff
@@ -0,0 +1,643 @@
+                                   as.character(counties)) %>%
+            as.factor() %>%
+            as.integer()
+        if (any(component > 1)) {
+            cli::cli_warn("Counties were not contiguous; expect additional splits.")
+        }
+    }
+
+
+
+    # get population stuff
+    pop_bounds <- attr(map, "pop_bounds")
+    pop <- map[[attr(map, "pop_col")]]
+    smallest_district_size <- attr(map, "district_seat_sizes") |> min()
+    if (any(pop >= pop_bounds[3] * smallest_district_size)) {
+        too_big <- as.character(which(pop >= pop_bounds[3] * smallest_district_size))
+        cli::cli_abort(c("Unit{?s} {too_big} ha{?ve/s/ve}
+                population larger than the district target.",
+                    "x" = "Redistricting impossible."))
+    }
+
+    # now check we don't have too many regions or counties
+    max_possible_sizes <- maximum_input_sizes()
+    if(dplyr::n_distinct(counties) > max_possible_sizes$max_counties ){
+        cli::cli_abort("The maximum number of supported counties is {max_possible_sizes$max_counties}!")
+    }
+    if(V > max_possible_sizes$max_V ){
+        cli::cli_abort("The maximum number of supported vertices in a map is {max_possible_sizes$max_V}!")
+    }
+    if(attr(map, "ndists") > max_possible_sizes$max_districts){
+        cli::cli_abort("The maximum number of supported districts in a map is {max_possible_sizes$max_districts}!")
+    }
+
+
+    return(list(
+        map=map,
+        adj_list=adj_list,
+        V=V,
+        ndists=ndists,
+        total_seats=total_seats,
+        district_seat_sizes=district_seat_sizes,
+        districting_scheme=districting_scheme,
+        counties=counties,
+        pop=pop,
+        pop_bounds=pop_bounds
+    ))
+
+}
+
+
+#' Checks that a constraint input is valid and returns it as a list ready for c++
+#'
+#' Checks that the sampling space/splitting method combination given is valid
+#' and that the splitting parameters are valid.
+#'
+#' @inheritParams redist_smc
+#'
+#'
+#' @returns A list of splitting parameters which is safe to pass into c++ code
+validate_constraints <- function(map, constraints, compactness = 1L){
+    # Other constraints
+    if (!inherits(constraints, "redist_constr")) {
+        constraints <- new_redist_constr(rlang::eval_tidy(rlang::enquo(constraints), map))
+    }
+    if (any(c("edges_removed", "log_st") %in% names(constraints))) {
+        cli::cli_warn(c("{.var edges_removed} or {.var log_st} constraint found in
+           {.arg constraints} and will be ignored.",
+                   ">" = "Adjust using {.arg compactness} instead."))
+    }
+    if (any(c("poslby", "fry_hold") %in% names(constraints)) && compactness == 1) {
+        cli::cli_warn("{.var polsby} or {.var fry_hold} constraint found in {.arg constraints}
+                 with {.arg compactness == 1). This may disrupt efficient sampling.")
+    }
+    constraints <- as.list(constraints) # drop data attribute
+    return(constraints)
```

Comment:
no trailing `return()`

I'll stop adding comments about this but just ctrl-F for instances of this and remove